File: 2603. Collect Coins in a Tree/Solution.java
// credit to @ye15

class Solution {
    public int collectTheCoins(int[] coins, int[][] edges) {
        int n = coins.length;
        Set<Integer>[] graph = new Set[n];
        Queue<Integer> leaves = new LinkedList<>();
        for (int i = 0; i < n; i++) graph[i] = new HashSet<>();
        for (int[] e : edges) {
            graph[e[0]].add(e[1]);
            graph[e[1]].add(e[0]);
        }
        for (int i = 0; i < n; i++) {
            int curr = i;
            while (graph[curr].size() == 1 && coins[curr] == 0) {
                int v = graph[curr].iterator().next();
                graph[curr].remove(v);
                graph[v].remove(curr);
                curr = v;
            }
            if (graph[curr].size() == 1) leaves.offer(curr);
        }
        for (int i = 0; i < 2; ++i) {
            for (int j = leaves.size(); j > 0; --j) {
                int curr = leaves.poll();
                if (graph[curr].size() > 0) {
                    int v = graph[curr].iterator().next();
                    graph[curr].remove(v);
                    graph[v].remove(curr);
                    if (graph[v].size() == 1) leaves.offer(v);
                }
            }
        }
        int s = 0;
        for (int i = 0; i < n; i++) s += graph[i].size();
        return s;
    }
}

File: 2603. Collect Coins in a Tree/Solution.py
# credit to @ye15

class Solution:
    def collectTheCoins(self, coins: List[int], edges: List[List[int]]) -> int:
        n = len(coins)
        graph = [set() for i in range(n)]
        leaves = deque()
        for x, y in edges:
            graph[x].add(y)
            graph[y].add(x)
        for i in range(n):
            while len(graph[i]) == 1 and coins[i] == 0:
                j = graph[i].pop()
                graph[j].remove(i)
                i = j
            if len(graph[i]) == 1: leaves.append(i)
        for i in range(2):
            for j in range(len(leaves)):
                curr = leaves.popleft()
                if graph[curr]:
                    v = graph[curr].pop()
                    graph[v].remove(curr)
                    if len(graph[v]) == 1: leaves.append(v)
        return sum([len(g) for g in graph])